### PR TITLE
Convert to use Get/Set-Timezone Cmdlets if available - Fixes #24

### LINKED ISSUE #24: Failing on Nano
VERBOSE: [NANOWINPE001]: LCM:  [ Start  Set      ]  [[xTimeZone]TimeZoneExample]
VERBOSE: [NANOWINPE001]:                            [[xTimeZone]TimeZoneExample]
 Performing the operation "Replace the System Time Zone" on target "'Dateline St
andard Time'".
VERBOSE: [NANOWINPE001]:                            [[xTimeZone]TimeZoneExample]
 Setting the TimeZone.
(3) : The type or namespace name 'Linq' does not exist in the namespace 
'System' (are you missing an assembly reference?)
(2) : using System.Collections.Generic;
(3) : >>> using System.Linq;
(4) : using System.Text;
    + CategoryInfo          : InvalidData: (Microsoft.Power...peCompilerError: 
   ) [], CimException
    + FullyQualifiedErrorId : SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands. 
   AddTypeCommand
    + PSComputerName        : localhost
 
(210) : 'Registry' is inaccessible due to its protection level
(209) :         {
(210) : >>>             var regTimeZones = 
Registry.LocalMachine.OpenSubKey("SOFTWARE\\Microsoft\\Windows 
NT\\CurrentVersion\\Time Zones");
(211) :             // Print out all the possible time-zones.
    + CategoryInfo          : InvalidData: (Microsoft.Power...peCompilerError: 
   ) [], CimException
    + FullyQualifiedErrorId : SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands. 
   AddTypeCommand
    + PSComputerName        : localhost
 
(216) : 'RegistryKey.GetSubKeyNames()' is inaccessible due to its protection 
level
(215) :             //}
(216) : >>>             var subKey = regTimeZones.GetSubKeyNames().Where(s => 
s == name).First();
(217) :             string daylightName = 
(string)regTimeZones.OpenSubKey(subKey).GetValue("Dlt");
    + CategoryInfo          : InvalidData: (Microsoft.Power...peCompilerError: 
   ) [], CimException
    + FullyQualifiedErrorId : SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands. 
   AddTypeCommand
    + PSComputerName        : localhost
 
(217) : 'RegistryKey.OpenSubKey(string, bool)' is inaccessible due to its 
protection level
(216) :             var subKey = regTimeZones.GetSubKeyNames().Where(s => s == 
name).First();
(217) : >>>             string daylightName = 
(string)regTimeZones.OpenSubKey(subKey).GetValue("Dlt");
(218) :             string standardName = 
(string)regTimeZones.OpenSubKey(subKey).GetValue("Std");
    + CategoryInfo          : InvalidData: (Microsoft.Power...peCompilerError: 
   ) [], CimException
    + FullyQualifiedErrorId : SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands. 
   AddTypeCommand
    + PSComputerName        : localhost
 
(218) : 'RegistryKey.OpenSubKey(string, bool)' is inaccessible due to its 
protection level
(217) :             string daylightName = 
(string)regTimeZones.OpenSubKey(subKey).GetValue("Dlt");
(218) : >>>             string standardName = 
(string)regTimeZones.OpenSubKey(subKey).GetValue("Std");
(219) :             byte[] tzi = 
(byte[])regTimeZones.OpenSubKey(subKey).GetValue("TZI");
    + CategoryInfo          : InvalidData: (Microsoft.Power...peCompilerError: 
   ) [], CimException
    + FullyQualifiedErrorId : SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands. 
   AddTypeCommand
    + PSComputerName        : localhost
 
(219) : 'RegistryKey.OpenSubKey(string, bool)' is inaccessible due to its 
protection level
(218) :             string standardName = 
(string)regTimeZones.OpenSubKey(subKey).GetValue("Std");
(219) : >>>             byte[] tzi = 
(byte[])regTimeZones.OpenSubKey(subKey).GetValue("TZI");
(220) :             var regTzi = new RegistryTimeZoneInformation(tzi);
    + CategoryInfo          : InvalidData: (Microsoft.Power...peCompilerError: 
   ) [], CimException
    + FullyQualifiedErrorId : SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands. 
   AddTypeCommand
    + PSComputerName        : localhost
 
Cannot add type. Compilation errors occurred.
    + CategoryInfo          : InvalidData: (:) [], CimException
    + FullyQualifiedErrorId : COMPILER_ERRORS,Microsoft.PowerShell.Commands.Ad 
   dTypeCommand
    + PSComputerName        : localhost
 
Unable to find type [Microsoft.PowerShell.xTimeZone.TimeZone].
    + CategoryInfo          : InvalidOperation: (Microsoft.PowerShell.xTimeZon 
   e.TimeZone:) [], CimException
    + FullyQualifiedErrorId : TypeNotFound
    + PSComputerName        : localhost
 
VERBOSE: [NANOWINPE001]: LCM:  [ End    Set      ]  [[xTimeZone]TimeZoneExample]
  in 11.6960 seconds.
The PowerShell DSC resource '[xTimeZone]TimeZoneExample' with SourceInfo 
'C:\share\SetTimeZone.ps1::21::9::xTimeZone' threw one or more non-terminating 
errors while running the Set-TargetResource functionality. These errors are 
logged to the ETW channel called Microsoft-Windows-DSC/Operational. Refer to 
this channel for more details.
    + CategoryInfo          : InvalidOperation: (:) [], CimException
    + FullyQualifiedErrorId : NonTerminatingErrorFromProvider
    + PSComputerName        : localhost
 
The SendConfigurationApply function did not succeed.
    + CategoryInfo          : NotSpecified: (root/Microsoft/...gurationManager 
   :String) [], CimException
    + FullyQualifiedErrorId : MI RESULT 1
    + PSComputerName        : localhost
 
VERBOSE: Operation 'Invoke CimMethod' complete.
VERBOSE: Time taken for configuration job to complete is 22.477 seconds

### PR DESCRIPTION
**Pull request by PlagueHO**
I also cleaned up the code a lot so that the variable names weren't being mixed up (sometimes TimeZoneId was used and sometimes it wasn't - the code was difficult to follow and debug).

I also improved the unit tests and increased the coverage.

I also performed a clean up of the code and bought it all up to HQRM standard. So hopefully I've eliminated the technical debt on this resource.

Here is a list of the changes:
* Add support for Nano Server and WMF5.1 via Get-Timezone/Set-Timezone cmdlets.
* Minor changes to bring make resource ready for HQRM.
* Renamed and reworked functions in TimezoneHelper.psm1 to prevent conflicts with new built-in WMF5.1 Timezone Cmdlets.
* Fixed localization so that failback to en-US if culture specific language files not available.
* Moved code to init C# type into Set-TimeZoneUsingNET functions
* Renamed internal Timezone parameters to TimezoneId to more clearly represent value

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/xtimezone/25)

### CODE REVIEW

File: DSCResources/TimezoneHelper.psm1 [hunk 1]
```diff
@@ -30,113 +38,146 @@ function New-TerminatingError
...
     else
     {
-        # For Nano Server the tzutil must be used
-        Write-Verbose -Message ($LocalizedData.SettingTimezoneTzUtilMessage `
-            -f $TimeZone)
-
-        try
+        $AddTypeCommand = Get-Command `
+            -Module Microsoft.Powershell.Utility `
+            -Name Add-Type `
```
Comment (Indhukrishna) [inline DSCResources/TimezoneHelper.psm1]: This block exists to load a particular module (and the cmdlet). Similar block exists for Timezone cmdlets also. Can we refactor this into a utility method so all the places can call this method instead?

**Comment by PlagueHO** [comment]
Thanks @Indhukrishna - good suggestion - done! Moved code to a Test-Command function in TimezoneHelpers.psm1 to detect if a CmdLet exists.

Are you able to use reviewable (click the Reviewable button in the PR) for any future review comments? Thank you heaps for the review.

**Comment by Indhukrishna** [comment]
Reviewed 5 of 8 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xtimezone/25)*

**Comment by Indhukrishna** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[DSCResources/TimezoneHelper.psm1, line 126 \[r1\]](https://reviewable.io:443/reviews/powershell/xtimezone/25#-KOOydwGVkP9JtkUybqG-r1-126:-KPE6liuhSE66lKPAOqV:bb6ywmx) ([raw file](https://github.com/powershell/xtimezone/blob/65aead32e9d40ee6892e92c19ed3a4d1f6617ceb/DSCResources/TimezoneHelper.psm1#L126)):*
<details><summary><i>Previously, Indhukrishna (Indhu Sivaramakrishnan) wrote…</i></summary>
> This block exists to load a particular module (and the cmdlet). Similar block exists for Timezone cmdlets also. Can we refactor this into a utility method so all the places can call this method instead? 
</details>
LGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xtimezone/25)*

**Pull request merged**